prompb: Add histograms to remote write/read protobufs

beorn7 opened the pull request:
_This is only for the sparsehistogram branch!_

A first attempt to add histograms to remote write/read protobufs.

@bboreham again cc'ing you to double-check if this makes sense.

codesome commented on prompb/types.proto:
```diff
@@ -54,13 +54,58 @@ message Exemplar {
+  ResetHint reset_hint      = 10;
+  // timestamp is in ms format, see model/timestamp/timestamp.go for
+  // conversion from time.Time to Prometheus timestamp.
+   int64 timestamp = 11;
```
```suggestion
  int64 timestamp = 11;
```

beorn7 replied:
Fixed.

codesome reviewed (approved):
With my limited knowledge on protos, LGTM

beorn7 commented:
Thanks, @codesome . Still waiting for remote write folks to comment.

Note that https://github.com/prometheus/client_model/pull/58 has changes to the _exposition format_ protobuf, adding support for float and gauge histograms, with some possibly enlightening comments.

bboreham commented on prompb/types.proto:
```diff
@@ -54,13 +54,58 @@ message Exemplar {
+// Sparse buckets.
+message Buckets {
+  message Span {
+    sint32 offset = 1; // Gap to previous span, or starting point for 1st span (which can be negative).
+    uint32 length = 2; // Length of consecutive buckets.
```
I note these are optional in `client_model`, and I couldn't figure out why.
If the answer is other than "mistake", please explain both cases.

beorn7 replied:
I _believe_ that's mostly historical. The proto spec in prometheus/client_model is using the proto2 syntax. Back then, protobuf had optional and non-optional fields. There were different schools of thought. Some considered `optional` fields harmful, some considered non-optional fields harmful. The Prometheus exposition protobuf was created following the latter school of thought, according to which you declare all fields `optional`, even if they aren't really optional.

proto3 resolved the schism by (arguably) declaring the "make everything `optional`" school as the winner. In proto3, all fields are implicitly optional. If they are missing, the default value is filled in, and (on the API level) you cannot actually distinguish between "unset" and "explicitly specified the value that happens to be the default value". If you really want an "unset" state, you have to declare a field as `repeated` but only care about zero or one repeats.

If the above is inaccurate, please pardon me. My context about protobuf is quite incomplete.

bboreham commented on prompb/types.proto:
```diff
@@ -54,13 +54,58 @@ message Exemplar {
+  repeated Span span = 1;
+  // Only one of "delta" or "count" may be used, the former for regular
+  // histograms with integer counts, the latter for float histograms.
+  repeated sint64 delta = 2; // Count delta of each bucket compared to previous one (or to zero for 1st bucket).
```
If I understand correctly, logically these values belong with the `Span`, but as an optimisation to reduce Protobuf metadata (tag and length) you have pulled them up one level and concatenated them, so the encoder/decoder must do a little extra work.
Either way, I think it would be worth some more comments so readers can understand.

beorn7 replied:
I think you are quite right. I'll add more comments.

For the decoder, it's actually less work, because the protobuf layout happens to be the same as the layout in TSDB. (Which isn't a reason to do it that way on its own, of course, just a convenient coincidence.)

beorn7 replied:
For remote write, it's actually the TSDB layout on both sides, so it's even less work for the encoder… (Encoding is more costly for exposition, depending on the way the instrumentation library manages the buckets internally.)

bboreham reviewed (commented):
No objections from me.
I do think you should link to the design (which I think is [here](https://docs.google.com/document/d/1cLNv3aufPZb3fNfaJgdaRBZsInZKKIHo9E6HinJVbpM/edit#).

Also mention / link to the explanation for "float histograms".

beorn7 commented:
> Also mention / link to the explanation for "float histograms".

I think the already mentioned https://github.com/prometheus/client_model/pull/58 is a good starting point to understand the need for "float histograms". I expect the most common use case for remote write will be to send the result of recording rules to the remote storage. (If you `rate` a histogram, you inevitably get a float histogram as a result.)

beorn7 commented:
I added two more things:

1. Utilized `oneof` (a proto3 feature) to mark the various one-of fields more explicitly. (Note that `oneof` isn't compatible with `repeated`, so for `count` vs `delta` in `Buckets`, we still depend on the user to only use one or the other.)
2. Added the `HISTOGRAM` chunk encoding, so that this should in principle be ready for chunked remote-read. (Note that `STREAMED_XOR_CHUNKS` turns out to be a bad name. Given that the `Chunk` message has an `Encoding type` field, it was always ready to contain non-XOR chunks, so the name in the enum should be simply `STREAMED_CHUNKS`. Not sure about the implication of changing it now, so I left the old name in place.)

beorn7 commented:
I have now also added a link to the design doc in the doc comments, including a pointer to the concept of float histograms.

This is just for the sparsehistogram branch, and it got a few looks already. I'm merging it now, but we can still iterate on it. Whoever works on implementing the actual use of these proto messages (myself, @cstyan, @csmarchbanks, maybe even @LeviHarrison), feel free to come back to the proto spec and adjust it to needs and insights found later.
